src/firefly/js/ui/panel/TabPanel.jsx

```diff
@@ -156,71 +159,74 @@ Tabs.propTypes= {
     borderless: PropTypes.bool
 };
 
-Tabs.defaultProps= {
+TabsView.defaultProps= {
     defaultSelected: 0,
     useFlex: false,
     resizable: false,
     borderless: false
 };
 
 
-export class Tab extends PureComponent {
-    constructor(props) {
-        super(props);
-    }
+/**
+ * TabPanel with internal state
+ * State will be lost after unload.
+ */
+export const Tabs = React.memo( (props) => {
+    const {defaultSelected=0, onTabSelect} = props;
+
+    const [selectedIdx, setSelectedIdx] = useState(defaultSelected);
 
-    componentDidMount() {
-        const {selected, onSelect, id, name} = this.props;
-        if (selected) {
-            onSelect(id, name);
+    const onSelect = useCallback( (index,id,name) => {
+        if (index !== selectedIdx) {
+            setSelectedIdx(index);
+            onTabSelect && onTabSelect(index,id,name);
         }
-    }
+    });
+
+    return (<TabsView {...props} defaultSelected={selectedIdx} onTabSelect={onSelect} />);
+});
+
+Tabs.propTypes = TabsView.propTypes;
+Tabs.defaultProps = TabsView.defaultProps;
+
 
-    render () {
-        const {name, label, selected, onSelect, removable, onTabRemove, id, maxTitleWidth} = this.props;
+/*----------------------------------------------------------------------------------------------*/
+/**
+ * TabPanel with ComponentCntlr supported state
+ * Selected state is stored as <componentKey>.selectedIdx
+ */
+export const StatefulTabs = React.memo( (props) => {
+    const {children=[], defaultSelected=0, onTabSelect, componentKey} = props;
+
+    let [selectedIdx = defaultSelected] = useStoreConnector( () => {
+        return get(getComponentState(componentKey), 'selectedIdx');
+    });
```

Comment:
I have a hard  time understanding what's happening here. We want selectedIdx state to be reset on every store update? Why not explicit useState here and set it in onSelect?

Reply:
Oh, it's because this will be a common use case.  That's why I created `useStoreConnector`.  This is a convenience way to listen/react to store changes.  This one function allow you to use one or many states connected to the `reduc store` and help cleanup the listener once unmount.

Reply:
I think the case you're talking about is `Tabs`, line `ui/panel/TabPanel.jsx:175`.  It uses internal state.. with useState.  Does this answer your question?

Reply:
Not quite. I though using store connector might be an overkill in this case, because you can set selectedIdx state in onSelect below, right after you do dispatchComponentStateChange. Because this is the only place selectedIdx should be changed. Or I am missing something?